## TOC:

*[01. Importing libraries]

*[02. Importing data]

*[03. Task]

# 01. Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Importing data

In [2]:
# defining path

path = r'C:\Users\Jonathan\08-2023 Instacart Basket Analysis'

In [3]:
# importing data sets

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# checking for null values in set

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [5]:
# identifying null data lines

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [6]:
#checking result

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [7]:
#checking dataframe shape

df_prods.shape

(49693, 5)

In [8]:
# removing null data values from data set

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [9]:
#checking new dataframe shape

df_prods_clean.shape

(49677, 5)

In [10]:
# identifying duplicate values in data set

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [11]:
#checking duplicate values

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [12]:
# creating new data set where duplicates are dropped

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [13]:
#checking new dataframe shape

df_prods_clean_no_dups.shape

(49672, 5)

In [14]:
#exporting dataframe

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 03. Task

In [15]:
# 2. Run df.describe() on df_ords dataframe. Share in a markdown cell whether anything about the data looks off or should be investigated further

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

2. order_day_of_week min and max is 0 and 6 but should be 1 and 7 

In [19]:
# 3. Check for mixed-type data in your df_ords dataframe

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [ ]:
# 4. If you find mixed-type-data, fix it.

# There was no mixed-type data found;

In [20]:
# 5. Run a check for missing values in your df_ords dataframe

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

5. There are 206,209 missing values in the days_since_last_order column; Further investigation to be conducted on why there are missing values;

In [25]:
#Create a df view for all missing values

df_ords_missing = df_ords[df_ords['days_since_last_order'].isnull() == True]

In [26]:
df_ords_missing.head(100)

Unnamed: 0  order_id  user_id eval_set  order_number  order_day_of_week  \
0              0   2539329        1    prior             1                  2   
11            11   2168274        2    prior             1                  2   
26            26   1374495        3    prior             1                  1   
39            39   3343014        4    prior             1                  6   
45            45   2717275        5    prior             1                  3   
...          ...       ...      ...      ...           ...                ...   
1455        1455   2242366       96    prior             1                  4   
1472        1472   1935233       97    prior             1                  4   
1478        1478   3131227       98    prior             1                  0   
1493        1493   2736359       99    prior             1                  4   
1512        1512    680467      100    prior             1                  3   

      order_hour_of_day  days_since_last_order  
0                     8                    NaN  
11                   11                    NaN  
26                   14                    NaN  
39                   11                    NaN  
45                   12                    NaN  
...                 ...                    ...  
1455                 12                    NaN  
1472                 15                    NaN  
1478                 14                    NaN  
1493                 15                    NaN  
1512                 18                    NaN  

[100 rows x 8 columns]

5. (cont.) The orders with NaN (missing values) in the days_since_last_order column seem to be a customers 1st order making it impossible for there to be a value for this column.

In [ ]:
# 6. Address the missing values using an appropriate method

# The solution for the missing values would be to add a column with string value where it is noted that it is a new customer;
# At this time I am unable to decipher how to achieve this but my best guess would be as follows;

# df_ords_new_customer = [df_ords["New Customer"] = ['Yes' if ('days_since_last_order'==NaN)= True, 'No' if ('days_since_last_order'==NaN)= False]];

In [27]:
# 7. Run a check for duplicate values in your df_ords data;

df_ords_dups = df_ords[df_ords.duplicated()]

7. No duplicates were found;

In [ ]:
#8. Address the duplicates using an appropriate method.

8. No duplicates were found;

In [ ]:
#9. Export your final, cleaned df_prods and df_ords data as ".csv" files in your "Prepared Data" folder and give them appropriate names.

In [28]:
#exporting dataframe

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))